nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/security/ACE.java

```diff
@@ -69,11 +71,24 @@ public static ACE fromId(String aceId) {
             return null;
         }
 
-        String[] parts = aceId.split(":");
-        if (parts.length < 3) {
-            throw new IllegalArgumentException(String.format("Invalid ACE id: %s", aceId));
+	// An ACE is composed of tokens separated with ":" caracter
+	// First 3 tokens are mandatory; following 3 tokens are optional
+        ArrayList<String> tokenList = new ArrayList<>();
+	Pattern p = Pattern.compile("^(.+):([^:]+):([^:]+):([^:]*):([^:]*):([^:]*)$");
+        Matcher m = p.matcher(aceId);
+        boolean b = m.matches();
+        if(!b) {
+	   throw new IllegalArgumentException(String.format("Invalid ACE id: %s", aceId));
+	}
+
+	for(int i=1; i <= m.groupCount(); i++) {
+		tokenList.add(m.group(i));
         }
 
+	// Turn ArrayList into an array of Strings	
+	String[] parts = new String[tokenList.size()];
```

Comment:
Allocate the `String[] parts` directly, no need for an intermediate `ArrayList<String> tokenList`.